### PR TITLE
Make WebUI and API code cleaner (+ 1.5 fixes)

### PR DESCRIPTION
**Pull request by Picus303**
This PR aims to make the code of the WebUI and API cleaner by spliting them into multiple files instead of having one giant file for each.

There is now a structured folder `tools/webui` that contains the same code as before but in a more readable way.

I'd like this PR to be the first of a many that'll aim to make the project easier to maintain.

**IMPORTANT:** This PR is not complete yet and only contains the new code for the WebUI. The API is probably broken right now. If anyone have a few samples of code that I could run to test it, I would be glad :)
The new code for the API should be ready quickly (I hope).

Personal note: I'd really like to help this project, but the giant files and use of global variables everywhere make it really difficult. Please, if it's possible, help me to help you by spliting your files :)

### CODE REVIEW

**Comment by Picus303** [comment]
btw, can someone please tell me why there are two files `api.py` and `postapi.py` ?
I want to be sure to understand.

**Comment by AnyaCoder** [comment]
> btw, can someone please tell me why there are three files `api.py`, `postapi.py` and `msgpack_api.py` ? I want to be sure to understand.

`api.py`: Server
`post_api.py`: using `requests` module to make a post request
`msgpack_api.py`: using `ormsgpack` module to make a post request

**Comment by Picus303** [comment]
> > btw, can someone please tell me why there are three files `api.py`, `postapi.py` and `msgpack_api.py` ? I want to be sure to understand.
> 
> `api.py`: Server `post_api.py`: using `requests` module to make a post request `msgpack_api.py`: using `ormsgpack` module to make a post request

Why do you need both?

**Comment by AnyaCoder** [comment]
> > > btw, can someone please tell me why there are three files `api.py`, `postapi.py` and `msgpack_api.py` ? I want to be sure to understand.
> > 
> > 
> > `api.py`: Server `post_api.py`: using `requests` module to make a post request `msgpack_api.py`: using `ormsgpack` module to make a post request
> 
> Why do you need both?

Well, at that time we just want to offer some examples to introduce how to handle `ormsgpack` api.
You can just remove `msgpack_api.py` now.

**Comment by Picus303** [comment]
Almost done! It seems to be working... There's just the code for the agent which is a nightmare to deal with.
I'll try to finish it this friday/saturday (timezones).

Technically, nothing changed to the way things work but the structure is very different. Simplified, shorter files and less repeated code. But it's a big PR, so if you want me to write explanations or have a call on discord, I understand.

**Comment by JJack27** [comment]
Wonder if we can customize the `temperature` and `top_p` parameter for web server?

**Comment by Picus303** [comment]
> Wonder if we can customize the `temperature` and `top_p` parameter for web server?

- If you mean the WebUI: Yes, from the UI
- If you mean the API: Also Yes, in your request

**Comment by Picus303** [comment]
I need to ask about the agent: this thing is a bit confusing.

I understand that internally, the agent uses the API to work, that's ok for me BUT:
2 of the 6 API endpoints `/v1/vqgan/encode` and `/v1/vqgan/decode` only makes sense in the context of the agent, and the endpoints for ASR and Chat are useful but are you sure that they should be directly included in this repo? The audio generation/finetuning part is already quite complex by itself.

It isn't the purpose of this PR, but I wonder if it wouldn't be better to give the agent it's own repo to better separate responsabilities.

**Comment by AnyaCoder** [comment]
> I need to ask about the agent: this thing is a bit confusing.
> 
> I understand that internally, the agent uses the API to work, that's ok for me BUT: 2 of the 6 API endpoints `/v1/vqgan/encode` and `/v1/vqgan/decode` only makes sense in the context of the agent, and the endpoints for ASR and Chat are useful but are you sure that they should be directly included in this repo? The audio generation/finetuning part is already quite complex by itself.
> 
> It isn't the purpose of this PR, but I wonder if it wouldn't be better to give the agent it's own repo to better separate responsabilities.

agent and tts use the same llama and vqgan, so we put them in one piece for convenience.

But this is a real problem and can lead to a codebase that is too complex. 

You need to keep those for now, we'll consider whether we want to move the agent code elsewhere later.

**Comment by AnyaCoder** [comment]
And pls see my review on your code(
If there are no more changes, please let me know so I can merge your code.

**Comment by Picus303** [comment]
> And pls see my review on your code( If there are no more changes, please let me know so I can merge your code.

I can't see it yet, but thanks, I'll have a look at it.
I'm currently working on the last changes to reimplement the chat endpoint in the API.

**Comment by Picus303** [comment]
And it's all done!
About the tests I have done, everything worked fine except one thing: loading the tokenizer which I guess comes from the changes between 1.4 and 1.5 as I reused exactly the same code.

`ValueError: The checkpoint you are trying to load has model type `dual_ar` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date.`

@AnyaCoder I still can't see your review, but if you're ok with what I wrote above, the code is ready to merge!

**Comment by AnyaCoder** [comment]
Masterpiece level refactor!

**Review by AnyaCoder — APPROVED** [review]

File: tools/run_webui.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    args.precision = torch.half if args.half else torch.bfloat16
+
+    # Check if CUDA is available
+    if not torch.cuda.is_available():
+        logger.info("CUDA is not available, running on CPU.")
```
[the remark this answers is not in the document]
Comment (teneous) [reply tools/run_webui.py:50]: before line 49.
```
# Check if mps is available
if torch.backends.mps.is_available():
    self.device = "mps"
    logger.info("mps is available, running on mps.")
```

**Comment by teneous** [comment]
I found that the code for reference audio is affected, and the `self.encode_reference function` in `reference_loader.py`  is not found.

**Comment by teneous** [comment]
> I found that the code for reference audio is affected, and the `self.encode_reference function` in `reference_loader.py` is not found.

```
 prompt_tokens = [
    self.encode_reference(
    # remove this line, no decoder_model args in method `encode_reference`
    # decoder_model=self.decoder_model,
    reference_audio=audio_to_bytes(str(ref_audio)),
    enable_reference_audio=True,
    )
    for ref_audio in ref_audios
]
```

**Comment by chaoqunxie** [comment]
Reference ID 填空会报错 TypeError: VQManager.encode_reference() got an unexpected keyword argument 'decoder_model' 填写了Reference ID 参考音频不生效

**Comment by chaoqunxie** [comment]
i use docker  fishaudio/fish-speech:latest

**Comment by teneous** [comment]
> Reference ID 填空会报错 TypeError: VQManager.encode_reference() got an unexpected keyword argument 'decoder_model' 填写了Reference ID 参考音频不生效

This Bug has already fixed in last weekend, use latest code to build a new docker image with branch main

**Comment by Jimwaa** [comment]
> And it's all done! About the tests I have done, everything worked fine except one thing: loading the tokenizer which I guess comes from the changes between 1.4 and 1.5 as I reused exactly the same code.
> 
> `ValueError: The checkpoint you are trying to load has model type `dual_ar` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date.`
> 
> Edit: Lengyue confirm it's normal.
> 
> @AnyaCoder I still can't see your review, but if you're ok with what I wrote above, the code is ready to merge!

`ValueError: The checkpoint you are trying to load has model type `dual_ar` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date.`
I have also encountered this problem, have you solved it?

**Comment by Picus303** [comment]
> > And it's all done! About the tests I have done, everything worked fine except one thing: loading the tokenizer which I guess comes from the changes between 1.4 and 1.5 as I reused exactly the same code.
> > `ValueError: The checkpoint you are trying to load has model type `dual_ar` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date.`
> > Edit: Lengyue confirm it's normal.
> > @AnyaCoder I still can't see your review, but if you're ok with what I wrote above, the code is ready to merge!
> 
> `ValueError: The checkpoint you are trying to load has model type `dual_ar` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date.` I have also encountered this problem, have you solved it?

That's something only the official team can solve, we have to wait.

**Pull request merged**